### PR TITLE
fix(dart): resolve a construction receiver reached through an import prefix (#2033)

### LINKED ISSUE #2033: Dart: an import-prefixed generic construction (`outer.Box<int>(1).height`) reassembles as `outer().height`, dropping the class hop
## Symptom

```dart
int f() { return outer.Box<int>(1).height; }
```

`dart_member_read_name` returns `'outer().height'`: the `.Box` hop is dropped, so the read names a member on whatever `outer()` resolves to rather than on `Box`.

Measured on the #2032 branch (1bcc3a75) with a direct parser probe; the unprefixed forms on that branch are correct (`Box<int>(1).height` -> `'Box().height'`).

## Mechanism

Same grammar mis-parse family as #2015: `Box<int>(1)` has no construction node, because `<` and `>` read as comparisons and the call's parens split off as a sibling `parenthesized_expression`. #2032 handles the unprefixed shape by taking the identifier leading the `X < int` left operand. With a prefix, that left operand is `outer.Box < int`, whose leading identifier is the PREFIX, not the class, so the class hop is lost.

## Expected

The prefixed form names the member on the constructed class, carrying the prefix the same way an unprefixed construction carries none: `outer.Box().height`, resolving through the import.

## Provenance

Found by the local review of #2032 and confirmed with a parser probe. Left out of that PR deliberately: it is a distinct receiver shape, and the fix there is already carrying a shadow-suppression guard for the same ambiguity.

No wrong edge was observed in the probe only because the prefix was untyped; the reassembled chain is wrong regardless, so a typed prefix could bind a same-named member of the prefix's own type.

### PR DESCRIPTION
**Pull request by vitali87**
Closes #2033.

## The issue was narrower than the bug

I filed #2033 from #2032's review saying the generic form dropped the class hop. Measuring on merged `main` first showed that was one of three causes, and not the one that mattered most:

```
p.Box<int>(1).height   -> 'p().height'      # hop dropped
new p.Box(1).height    -> 'p().height'      # hop dropped (not in the issue)
p.Box(1).height        -> 'p.Box().height'  # name CORRECT, still no edge
```

The bare form produced a perfectly good name and emitted nothing, so a chain-walker-only fix would have passed a narrow test and left the common case broken.

**Three causes**, each mutated individually and each reddening its own test:

1. **The prefix was never registered.** `_parse_dart_imports` keyed the mapping on `dart_local_name(uri)` — the *file* name — so `import 'lib.dart' as p;` mapped `lib` and never `p`, the name the source actually uses.
2. **The resolver ended the chain at the prefix.** `_infer_chained_object_type` looked `p` up in `local_var_types`, which an import prefix is not, returning None before the construction hop.
3. **The walker took the prefix for the class** in both construction shapes.

## Three defects found in my own fix

Each was found by probing the change rather than by a test failing, and each is now pinned:

| Defect | Why it mattered |
|---|---|
| The fold gated on "registry knows this name", and the registry holds **functions** too, so `mod2.factory().run()` bound a same-named *class* in an unrelated module | `main` emitted **no** edge there — I replaced silence with a confidently wrong edge pointing at the wrong module |
| The alias was written unconditionally, so `import 'helper.dart'; import 'other.dart' as helper;` **dropped the first import entirely** | Both IMPORTS edges pointed at `other`, and unprefixed `H()` stopped resolving |
| `dart_ambiguous_construction_base` now returns a dotted `p.Box`, but the shadow check keys on the bare binder, so the lookup could never match | A local named `p` stopped suppressing `p.Box < b > (1).x` — reintroducing the wrong-edge class #2015's shadow check exists to prevent |

## Cross-language impact, measured not assumed

`_infer_chained_object_type` is language-agnostic, so this changes more than Dart. Measured against `origin/main` in a synced worktree (15 parsers, not the 1-parser default that silently makes every such comparison look clean):

- **Python** gains a correct edge: `mod.Cls().method()` emitted only `INSTANTIATES` before, now binds `mod.Cls.method` and avoids a same-named decoy. Pinned by its own test, since an undeclared cross-language change in a Dart-titled PR is how silent regressions ship.
- **Go** unchanged (already resolved this shape via its own path). **JS** unaffected.

## Verified by construction, not by sampling

A `show`/`hide` combinator cannot be mistaken for a prefix: its identifier is nested inside a `combinator` node, while the `as` prefix is a direct child of `import_specification`, and the extractor scans only direct children. `export`, `deferred as`, and `as p show Foo` all behave correctly. A generic return type (`List<Thing>`) declines rather than binding a bogus member; a nullable one (`Thing?`) binds correctly.

## Tests

Affected suites: **998 passed, 0 failed**, 30 errors all `@memgraph` (the documented Docker-less baseline on this host).

## Process note

The `greptile-local` reviewer stalled twice on this branch and was killed both times, so the risks it was asked about were verified directly instead — that is how the second and third defects above were found. Flagging it rather than claiming a clean review round.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved Dart analysis for import-prefixed constructors and chained getter references.
  * Correctly handles named constructors and local variables that shadow import prefixes, reducing false references.
  * Preserved accurate import relationships when aliases overlap.
  * Improved Python resolution for module-qualified class construction and factory calls.

* **Tests**
  * Added coverage for Dart import prefixes, getter references, alias conflicts, and comparison expressions.
  * Added coverage for Python module-qualified method and factory calls.

### CODE REVIEW

**Comment by vitali87** [comment]
Follow-up from a bounded read-only review of the diff, pushed as 57a7140b. Three of its findings acted on, two dismissed by execution.

**Acted on.**

The function-hop branch had no POSITIVE test. The factory guard test asserts only an absence, which a fold-nothing implementation satisfies equally — assertion-satisfied-by-both — so the review reasonably asked whether the branch was dead. It is not: removing it drops the `REFERENCES` edge from `p.make().v` to `Thing.v`. Measured, then pinned with a test that goes red without the branch. Worth stating plainly because the suggested remedy was to delete it, which would have removed working behaviour.

Two docstrings claimed more than their code. `dart_ambiguous_construction_base` still said 'the leading identifier' after it began returning a dotted name under a prefix — precisely the mismatch that caused the shadow-check defect fixed in 99e4b0ea — and `dart_import_prefix` claimed a plain/`deferred as` distinction the code does not make. Both now describe what the code actually does.

**Dismissed by execution.**

The review's highest-priority unverified item was whether the `new`/`const` comprehension could absorb argument identifiers into the dotted name (`new p.Box(someLocal)` -> `p.Box.someLocal`). It cannot: arguments sit in a distinct `arguments` node that the type filter excludes. Verified for one argument, two arguments, and the const form — all yield `p.Box().height`.

A prefix-qualified return type (`q.Thing`) does resolve to nothing, as the review predicted. I checked the direction that matters: it does NOT bind a same-named class elsewhere. Under-resolution, never mis-resolution, and a pre-existing limit of the shared chain path rather than something this PR introduces — so it is recorded rather than fixed here.

It also independently reached the shadow-check defect I had already found and fixed in 99e4b0ea, which is a useful confirmation that the fix was warranted.

Suites on the pushed head: 999 passed, 0 failed, 30 errors all `@memgraph`.

@greptileai review

**Comment by vitali87** [comment]
Two more real defects fixed in 953f6d70, both mine, both confirmed against `origin/main` in a synced worktree before I touched anything.

**A local shadowing the prefix still folded** (CodeRabbit's scope-awareness finding). A parameter named `p` makes `p.Box(1).height` a read on that parameter, but it resolved through `import ... as p` and emitted CALLS and REFERENCES edges. `main` emits none, so I had invented them. The fold now declines when the leading segment is a known local.

CodeRabbit was also right that typed and untyped bindings differ, and the fix does not cover both: an untyped Dart local (`var p = 1`) never reaches `local_var_types` — I instrumented the fold and it arrives empty — so the resolver cannot see it shadowing. That needs the call processor's shadow spans threaded into resolution, which is wider than this PR. It is recorded as a named gap in the test rather than quietly left.

**Named constructors were absorbed into the receiver type** (Greptile + CodeRabbit). `new Box.named(1)` became type `Box.named`, not a definition, so the read resolved against nothing — and `main` resolves it, so this was a regression I introduced, not a pre-existing gap.

Worth spelling out because my first two attempts at this were wrong: the grammar cannot settle it alone. After one `type_identifier` the trailing `identifier` is the CLASS in `new p.X(1)` and the CONSTRUCTOR in `new X.named(1)` — byte-identical shapes. Two `type_identifier`s DO settle it (prefix and class both named), so that case drops the constructor in the walker; the ambiguous pair is left to the resolver, which types a base naming a class directly. I only got there after measuring all five shapes; reasoning from two samples produced a fix that broke the case it was meant to preserve.

Each new guard reddens its own test alone. Affected suites on the pushed head: 1001 passed, 0 failed, 30 errors all `@memgraph`.

@greptileai review

**Comment by vitali87** [comment]
Correcting something I asserted on this PR: I said the resolver "genuinely cannot see" an untyped local shadowing an import prefix, and recorded it as a gap that needed shadow spans threaded from the call processor. Greptile's P1 (`Handle untyped prefix shadows`) is right, and my reason for not fixing it was wrong.

I had measured `local_var_types` and found `var p = 1` never arrives there. That part is true. But I concluded from it that the information is unavailable, without checking the syntactic path in the same file. It is available:

```
$ declared-name hits for `int untyped() { var p = 1; return p.Box(1).height; }`
[('initialized_variable_definition', ['p'])]
```

`_dart_declared_names` binds `p` from the declaration itself, with no type inference involved, and `_dart_shadow_spans` already turns those into byte spans — the same machinery the #2015 guard uses. So the shadow check does not need new analysis, it needs the existing spans reaching `_fold_import_prefix_hop`, which today takes only `local_var_types`.

That is a cross-module change (`call_processor` holds the spans, `call_resolver` makes the fold decision, and `_infer_chained_object_type` has one call site), so it is not a one-line patch, and I am not going to smuggle it into this PR's diff while the bots are anchored to the current head.

Two honest options, and I would rather you picked:

1. Fix it here. The diff grows by the parameter thread plus a test, and everything on this PR gets re-reviewed from a new head.
2. Land this PR for the two shadowing/named-constructor fixes it does make, and take the untyped case as a separate issue with the mechanism above written down.

I lean 1, because what ships otherwise emits false REFERENCES/CALLS edges for a real Dart shape, and "documented" does not make a wrong edge safe — that was the flaw in how I recorded it the first time.

The other two findings (`Honor explicit Dart aliases`, `Handle named constructors`) were both fixed in 953f6d70, which is the commit this review is anchored to. If they are still reproducing I would want the failing input, because my own checks say otherwise.

**Comment by vitali87** [comment]
Pushed `1c8608ef` for the two Sonar CRITICALs, plus a correction and a new issue that came out of verifying them.

**Sonar.** `Sonar Zero Issues Gate` was red with two `python:S3776` cognitive-complexity issues, both in functions this PR adds: `_construction_class_name` at 27 and `_fold_import_prefix_hop` at 19, limit 15. Note `SonarCloud Analysis` was green throughout — the Gate is the check that enforces zero issues, and the two disagree.

Both are pure extractions along seams that were already there:

- `_construction_class_name` handled two unrelated grammars in one body, so it splits into `_explicit_construction_name` (new/const) and `_mis_parsed_generic_name` (the relational shape).
- `_fold_import_prefix_hop` hands its named-constructor resolution to `_drop_named_constructor`.

40 Dart and Python resolution tests pass unchanged.

**A pre-existing bug the refactor exposed, now #2084.** I mutated `_drop_named_constructor` to a no-op expecting the named-constructor test to redden. All 28 tests stayed green. The cause is that Dart's two spellings take different code paths:

```
new p.Box.named(1)  ->  new_expression(type_identifier p, type_identifier Box, identifier named, arguments)
p.Box.named(1)      ->  identifier(p) + selector(.Box) + selector(.named) + selector((1))   <- no construction node
```

Every fixture in the repo writes `new` or `const`, so the fold path had no coverage at all. Writing the missing test showed the bare spelling — the idiomatic one since Dart 2 — emits no edge whatsoever, and it fails identically on `origin/main` in a 15-parser worktree. Pre-existing, so I filed it as #2084 with the AST dump and the failing test rather than widening this diff.

**Correcting the known-gap comment.** It claimed the resolver "cannot see" an untyped local shadowing the prefix. That is wrong, as I noted in my previous comment: `_dart_declared_names` binds `var p = 1` syntactically, with no type inference involved. The comment now says so and names the shadow-span route, so nobody reads it as settled.

The untyped-local P1 itself is still open and still needs the owner's call on whether it lands here or separately — it emits false edges, so "documented" is not a resolution.

@greptileai review

**Comment by vitali87** [comment]
Pushed `9e074b88`, which supersedes the review I triggered on `1c8608ef` — please re-anchor to this head.

The first split was not enough. I checked the result instead of assuming it, and `_explicit_construction_name` came out sitting ON the limit rather than under it, which is not a margin: any later edit to it reopens the gate. Its child scan (collect the leading `type_identifier`s, then a trailing `identifier`) now lives in `_construction_name_parts`, leaving the caller holding only the positional rule that decides prefix vs named constructor.

That also removed a `return None` the first split had left unreachable after the final return.

Measured, against a limit of 15:

```
_construction_class_name       1
_explicit_construction_name    3
_construction_name_parts      12
_mis_parsed_generic_name      11
_drop_named_constructor        3
_fold_import_prefix_hop       14
```

The same 40 Dart and Python resolution tests pass on this head.

Worth being explicit that this is the second attempt at the same finding, because it is the same mistake in miniature that I made earlier on this PR: I reported the complexity fix as done after one split, having estimated rather than measured. The estimate and the limit happened to be the same number.

@greptileai review

**Comment by vitali87** [comment]
Pushed `b7100642`, which closes the untyped-prefix-shadow P1 you raised. Head moved twice since the last review, so please anchor to this one.

## The fix

The import processor now records, at parse time, the byte spans each import prefix is shadowed in, and the fold checks the call site against them. The recording is purely syntactic — `var p = 1` binds `p` whether or not a type can be inferred — which is precisely why it sees what `local_var_types` cannot.

Two things it needed beyond the guard itself, both found by measuring rather than reading:

- **The read path passed no position.** The guard saw `call_point=None` on every read and declined. `_emit_dart_property_read` now threads `node.start_byte` into `resolve_function_call`, whose `call_point` parameter already existed for Rust block scopes.
- **A parameter's span swallowed the whole file.** A parameter lives in the signature, a SIBLING of the body, so walking up for a block ancestor finds none and falls through to the root. Measured at `[0,121)` — it covered an unrelated read 45 bytes later and broke the control in the existing shadow test. Scoped to the body following its signature it is `[48,75)`.

## What the local review caught, and what I got wrong fixing it

**P1: pattern declarations were missing.** `var (p, q) = (1, 2)` recorded no span, reproducing the exact defect the guard exists to stop. The call processor's `_dart_declared_names` already walks patterns; my copy diverged and was wrong. Verified before fixing — mine returned `{}` where the existing one returned `['q', 'p']`.

**P2: the parameter rule was untested — and my first fix for it did not work either.** Deleting the signature branch left every test green. I wrote a test, and it stayed green too. Two reasons, both worth recording:

- the fixture used `var p = q` as the binder, and a local declaration HAS a block ancestor, so it never reaches the signature branch at all;
- a typed parameter is caught upstream by `local_var_types`, which masks any difference the span rule makes.

`int first(p)`, unannotated, is the only shape where the span path is the sole suppressor. With the rule removed it leaks `first REFERENCES Box.height`. Now it reddens on exactly that test.

## A correction, and a new issue

My pattern test originally asserted `not INSTANTIATES`. That was over-reach: this fix does not deliver it. Measured at this commit's parent, the CALLS/INSTANTIATES pair leaks under **every** shadow shape — including the typed parameter the older guard does catch for reads — because the call path resolves through `_try_resolve_via_import` and never reaches this fold:

```
call_resolver:2909 - Import-resolved static call: p.Box -> proj.lib.Box
```

| shadow shape | CALLS | INSTANTIATES | REFERENCES |
|---|---|---|---|
| `dynamic p` parameter | yes | yes | no (guarded) |
| `var p = 1` | yes | yes | no (this PR) |
| `var (p, q) = (1, 2)` | yes | yes | no (this PR) |

Pre-existing and shape-independent, so it is **#2088** rather than scope creep here. The machinery to fix it already exists — `dart_prefix_shadows` plus `_dart_prefix_is_shadowed`, and the call path already threads `call_point`.

I briefly concluded this was a regression I had caused. It was not: I had compared against a worktree based on `origin/main`, which lacks the whole #2033 feature, so its clean result meant "feature absent", not "guard working". Re-measured against the actual parent commit.

## Verified

- 274 passed across the Dart, Python-resolution, call-resolver, call-processor and import-parsing suites; 373 earlier across sixteen symbol-selected suites covering C, C++, C#, PHP and Python — no cross-language regression from the `call_processor`/`import_processor` changes.
- Every fix line mutates red independently: the guard, the span recording, the pattern walk, the signature rule.
- Cognitive complexity kept under the limit by extracting `_names_own_constructible` and `_following_body_span` (fold 13, span walk 10, limit 15).
- `ty check` fails on this branch and on the unmodified branch alike; `ty` on the file this commit touches reports `All checks passed!`.

@greptileai review

File: codebase_rag/parsers/call_resolver.py [hunk 1]
```diff
@@ -3345,6 +3357,43 @@ def _infer_chained_object_type(
...
+        qn = f"{target}{cs.SEPARATOR_DOT}{name}"
+        kind = self.function_registry.get(qn)
+        if kind in _CONSTRUCTIBLE_NODE_TYPES:
+            return name, parts[1:]
+        if kind is None:
+            return None
+        # A function hop types the chain by its RECORDED return type instead;
+        # with none recorded the chain stays unresolved rather than guessing.
+        returned = self.type_inference.method_return_types.get(qn)
+        return (returned, parts[1:]) if returned else None
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/parsers/call_resolver.py]: Fixed in d73380db. The function hop now returns `_chain_class_qn(returned, target)`, which qualifies the bare return type in the factory's own module, so a same-named class in the caller's scope no longer wins. Test: `test_a_prefixed_factory_types_the_chain_in_its_own_module`. The decoy is `alt.dart` because it sorts before `lib`; the unfixed code bound `alt.Thing.v`.

File: codebase_rag/parsers/import_processor.py [hunk 2]
```diff
@@ -4615,6 +4628,32 @@ def _parse_dart_imports(self, captures: dict, module_qn: str) -> None:
...
+                # same file elsewhere in the module.
+                # An alias must not clobber a key another import already
+                # owns: `import 'helper.dart'; import 'other.dart' as helper;`
+                # would otherwise drop the first import entirely, since the
+                # file-derived key and the alias collide. The prefix is the
+                # name the source uses for THIS import, so it is only added
+                # where it is free.
+                prefix = dart_import_prefix(import_node)
+                if prefix and prefix not in self.import_mapping[module_qn]:
+                    self.import_mapping[module_qn][prefix] = full_name
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/parsers/import_processor.py]: Fixed in d73380db. When the prefix enters `import_mapping`, its site is recorded with `_record_import_site(module_qn, prefix, import_node, uri)`, so its IMPORTS edge carries the span and alias. Test: `test_a_dart_import_prefix_binding_carries_its_site`, which fails on the previous head.

File: codebase_rag/parsers/import_processor.py [hunk 3]
```diff
@@ -4615,6 +4628,32 @@ def _parse_dart_imports(self, captures: dict, module_qn: str) -> None:
...
+                    prefixes.add(prefix)
+        # A local or parameter of the same name SHADOWS the prefix inside its
+        # scope, and an UNTYPED one (`var p = 1`) never reaches the resolver's
+        # local_var_types, so the type map cannot answer this. The binding is
+        # syntactic, so record its span here and let the fold check the call
+        # site against it (issue #2033).
+        if root_node is not None and prefixes:
+            self.dart_prefix_shadows[module_qn] = dart_binding_spans(
+                root_node, frozenset(prefixes)
+            )
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/parsers/import_processor.py:4686]: Fixed in 03857e08. `_clear_module_import_state` pops `dart_prefix_shadows` and `dart_import_aliases` on every re-parse, so an empty-prefix parse leaves nothing behind. `test_a_reparse_drops_the_dart_prefix_state` covers it: removing the pops reddens it.

File: codebase_rag/parsers/import_processor.py [hunk 4]
```diff
@@ -4615,6 +4635,40 @@ def _parse_dart_imports(self, captures: dict, module_qn: str) -> None:
...
+                prefix = dart_import_prefix(import_node)
+                if prefix:
+                    # An explicit `as` prefix is the name the SOURCE uses for
+                    # this import, so it owns that name outright. Recorded in
+                    # its own map rather than overwriting import_mapping,
+                    # whose values carry the IMPORTS edges: writing it there
+                    # dropped the colliding unprefixed import entirely.
+                    self.dart_import_aliases.setdefault(module_qn, {})[prefix] = (
+                        full_name
+                    )
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/parsers/import_processor.py]: Fixed in d73380db. `dart_import_aliases` now keeps every library under a shared prefix. The fold searches all of them and binds only when exactly one library defines the name; an ambiguous import binds nothing. Tests: `test_libraries_sharing_a_prefix_are_all_searched` and `test_a_name_both_prefixed_libraries_define_is_not_guessed`, both failing on the previous head.
